### ISSUE TITLE
 db.exec() FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - process out of memory

### ISSUE THREAD

**Issue by bytebrahma** [orientechnologies/orientjs#148]
Hi,

I am using db.exec() to insert edges. Have created this as a method in my class orientServer:
orientServer.prototype.execSql=function(sql,params){
    var cur=this;
    //console.log(this.db);
    return new Promise((resolve,reject)=>{

```
    var paramObj={"params":params};
    //log.info(sql)
    cur.db.exec(sql,paramObj)
    .then(function(val){
        log.info("success.."+val)
        resolve(val);
    })
    .catch(function(err){
        //log.info("Error in execSql "+err);
        log.info("error.."+val)
        resolve(err);
    })
})
```

}

this function is being called in a loop going over 390071 elements' array from my main program to create Edges(using Promise.all). The resulting sql is:

```
create edge psrecfield from (SELECT FROM psrecd
efn where recname=:sourceVal) to (SELECT FROM psdbfield WHERE fieldname = :targe
tVal) content :contentObj
```

params object is:

```
{ params:
   { sourceVal: 'TL_RPTD_APP_VW2',
     targetVal: 'ORIGINATORID',
     contentObj:
      { RECNAME: 'TL_RPTD_APP_VW2',
        FIELDNAME: 'ORIGINATORID',
        EDITTABLE: ' ',
        USEEDIT: 8388608,
        SETCNTRLFLD: ' ',
        SUBRECORD: 'N' } } }
```

 The program runs for a while without printing success/error and terminates with the below error:

```

<--- Last few GCs --->

  372933 ms: Scavenge 1397.3 (1457.5) -> 1397.3 (1457.5) MB, 16.9 / 0 ms (+ 2.0
ms in 1 steps since last GC) [allocation failure] [incremental marking delaying
mark-sweep].
  374217 ms: Mark-sweep 1397.3 (1457.5) -> 1397.2 (1457.5) MB, 1283.3 / 0 ms (+
3.0 ms in 2 steps since start of marking, biggest step 2.0 ms) [last resort gc].

  375502 ms: Mark-sweep 1397.2 (1457.5) -> 1397.1 (1457.5) MB, 1285.1 / 0 ms [la
st resort gc].


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 00000238723B4639 <JS Object>
    1: derez(aka derez) [D:\dev\sputnik\node_modules\winston\node_modules\cycle\
cycle.js:90] [pc=000001D8A4ADB893] (this=00000238723041B9 <undefined>,value=0000
023872309D01 <an Object with map 000003762B0087D1>,path=0000023872390A21 <String
[1]: $>)
    2: decycle [D:\dev\sputnik\node_modules\winston\node_modules\cycle\cycle.js:
~24] [pc=000001D8A4BC5299] (this=00000167DBC595E1 <an Object with ...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - process out of memory

```

Also I can see that around 137 edges are inserted at the end. Next time I run the program the same 137 edges are added ignoring the index I have created. Could you please help me understand what is the issue? I am perplexed because the program successfully inserts around 110k vertices using db.class.Create();

**Comment by bytebrahma** [comment]
May be not an efficient way, but I got around this with Event emitters. Below is the code. i have stacked up all the promises as closures in an array and processing them in  steps using recursive events from the event emitter. Closures are such a wonderful concept in JS!! 

````
var events = require('events');
		var eventEmitter = new events.EventEmitter();
		
		
		eventEmitter.on('nextChunk',function(promises,st,step){
			console.log("promises,st,step=",promises.length,st,step)
			var max = promises.length;
			var end= Math.min(max,st+step);
			var exit=false;
			var pArr=[];
			
			if(end==max) exit=true;
			
			for(var i=st;i<end;i++)
			{
				pArr.push(promises[i]());
				
			}			
			
			Promise.all(pArr)
			.then((resolve,reject)=>{
				console.log("Promises handled till  ",end)
				if(!exit)
				 this.emit("nextChunk",promises,end,step);
				else resolve(end);
			})
			
			
		}) // END on nextChunk
	eventEmitter.emit('nextChunk',pArr,0,50000);
	}) // END Promise
````

**Comment by jarno-steeman** [comment]
Bluebird can help you out, check the `concurrency` option on their [Promise.map](http://bluebirdjs.com/docs/api/promise.map.html) function

**Comment by bytebrahma** [comment]
Thanks Jarno. That helped!

**Comment by maggiolo00** [comment]
hi @bytebrahma 

yes in general returning such a big results set could be a problem
see also here
https://github.com/orientechnologies/orientjs/issues/116
We kept that issue as enhancement to introduce streaming support.

**Comment by wolf4ood** [comment]
Streaming support is available in OrientJS 3.0.x

https://orientdb.com/docs/3.0.x/orientjs/Session.html#streaming
